### ISSUE TITLE
Type Assertion using `as` should warn about wrong types

### ISSUE THREAD

**Issue by rubenferreira97** [microsoft/TypeScript#50698]
# Bug Report

### 🔎 Search Terms
type assertion

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `as`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.0.5#code/JYOwLgpgTgZghgYwgAgCoQM5mQbwLABQyycAymFKAOYBcyWlIVANIQL6GEID2IWJAMQCuIBGGC8ATMgC8yABQBbYFQAWYAEIQAckMUAjaMjoNqyAD7IQew1ACUsgHy5CxHn2xx0-OThLlGWmRlNU0dGyM2Egw0TDAAbmQAeiTkUlVuIQAbABMQAHJsAHduKABrV2R3fi846TpvbF9-Cmo6EPUtXQNIxJTkAGFSqAgxZGgoUvZOAmrPY3pWpgsrCKhZZABmeK5efn0NvzIloLhkKLgYxr7UoagRsYmp2b3sBAXGw5bAujO2G8Gw1G2CeUCAA)

### 💻 Code

```ts
interface Test {
  aString: string,
}

const aFunction2 = (mightBeNumber : string | number) => {
  const aTest = { aString: mightBeNumber } as Test; // Shouldn't work
  const aTest2 : Test = { aString: mightBeNumber }; // Correct error
}

const a : string | number = 3;
// here typescript knows that `a` is always a number
const b = { aString: a } as Test; // Correct error
const c : Test = { aString: a }; // Correct error
```

### 🙁 Actual behavior

Type Assertion using `as` does not warn about wrong types

### 🙂 Expected behavior

Type Assertion using `as` should warn about wrong types

**Comment by MartinJohns** [comment]
The entire point of type assertions is to work around the type checks of the compiler. You're telling the compiler "trust me, it matches, or will match later one, so shaddap."

**Comment by rubenferreira97** [comment]
@MartinJohns Then I would assume that typescript shouldn't warn me about trying to convert a number to string like `const test = "a" as number;`

**Comment by MartinJohns** [comment]
It does perform some basic checks to make sure the types at least **somewhat** overlap. That means you can't say a string is a number, but you can say "a string or a number" is definitely a number.

**Comment by Josh-Cena** [comment]
> `as unknown/any as number`

That's a feature. When you write `"a" as number`, the error message literally tells you to cast to `unknown` first.
